// react-static/lib/static/components/Redirect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireDefault(require("react"));

var _reactHelmet = _interopRequireDefault(require("react-helmet"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

//
var Redirect = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(Redirect, _React$Component);

  var _super = _createSuper(Redirect);

  function Redirect() {
    (0, _classCallCheck2["default"])(this, Redirect);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(Redirect, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          to = _this$props.to,
          fromPath = _this$props.fromPath;
      var resolvedTo = (0, _typeof2["default"])(to) === 'object' ? to.pathname : to;

      if (!resolvedTo.includes('//')) {
        resolvedTo = "".concat(process.env.REACT_STATIC_PUBLIC_PATH).concat(resolvedTo === '/' ? '' : resolvedTo);
      }

      return /*#__PURE__*/_react["default"].createElement(_reactHelmet["default"], null, fromPath && /*#__PURE__*/_react["default"].createElement("title", null, "".concat(process.env.REACT_STATIC_PUBLIC_PATH).concat(fromPath === '/' ? '' : fromPath)), /*#__PURE__*/_react["default"].createElement("link", {
        rel: "canonical",
        href: resolvedTo
      }), /*#__PURE__*/_react["default"].createElement("meta", {
        name: "robots",
        content: "noindex"
      }), /*#__PURE__*/_react["default"].createElement("meta", {
        httpEquiv: "content-type",
        content: "text/html; charset=utf-8"
      }), /*#__PURE__*/_react["default"].createElement("meta", {
        httpEquiv: "refresh",
        content: "0; url=".concat(resolvedTo)
      }));
    }
  }]);
  return Redirect;
}(_react["default"].Component);

exports["default"] = Redirect;
