// react-static/lib/static/webpack/rules/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.getStagedRules = void 0;

var _jsLoader = _interopRequireDefault(require("./jsLoader"));

var _cssLoader = _interopRequireDefault(require("./cssLoader"));

var _fileLoader = _interopRequireDefault(require("./fileLoader"));

var _jsLoaderExternal = _interopRequireDefault(require("./jsLoaderExternal"));

var getStagedRules = function getStagedRules(args) {
  return {
    jsLoader: (0, _jsLoader["default"])(args),
    jsLoaderExt: (0, _jsLoaderExternal["default"])(args),
    cssLoader: (0, _cssLoader["default"])(args),
    fileLoader: (0, _fileLoader["default"])(args)
  };
};

exports.getStagedRules = getStagedRules;

var _default = function _default(args) {
  return [{
    oneOf: [(0, _jsLoader["default"])(args), (0, _jsLoaderExternal["default"])(args), (0, _cssLoader["default"])(args), (0, _fileLoader["default"])(args)]
  }];
};

exports["default"] = _default;
