// react-static/lib/index.node.js
"use strict";

var _require = require('../lib/commands/create'),
    create = _require["default"];

var _require2 = require('../lib/commands/start'),
    start = _require2["default"];

var _require3 = require('../lib/commands/build'),
    build = _require3["default"];

var _require4 = require('../lib/commands/bundle'),
    bundle = _require4["default"];

var _require5 = require('../lib/commands/export'),
    exporter = _require5["default"];

var _require6 = require('../lib/static/webpack/runDevServer'),
    reloadClientData = _require6.reloadClientData;

var _require7 = require('../lib/node/makePageRoutes'),
    makePageRoutes = _require7["default"];

var _require8 = require('../lib/node/getWebpackConfig'),
    getWebpackConfig = _require8["default"];

var _require9 = require('../lib/static/getConfig'),
    normalizeRoutes = _require9.normalizeRoutes;

var _require10 = require('../lib/static/createSharedData'),
    createSharedData = _require10["default"];

var _require11 = require('../lib/static/getRoutes'),
    rebuildRoutes = _require11.rebuildRoutes;

module.exports = {
  create: create,
  start: start,
  build: build,
  bundle: bundle,
  "export": exporter,
  reloadClientData: reloadClientData,
  makePageRoutes: makePageRoutes,
  getWebpackConfig: getWebpackConfig,
  normalizeRoutes: normalizeRoutes,
  createSharedData: createSharedData,
  rebuildRoutes: rebuildRoutes
};
